File: applications/webapp/src/core/port/driven/broadcast/room/dashboard/AdminDashboardBroadcastRoom.ts
import { BroadcastRoom } from "@/core/port/driven/broadcast/BroadcastRoom";
import { AnnouncementResponseDTO } from "@/core/port/dto/response/announcement/AnnouncementResponseDTO";
import { ClarificationResponseDTO } from "@/core/port/dto/response/clarification/ClarificationResponseDTO";
import { LeaderboardCellResponseDTO } from "@/core/port/dto/response/leaderboard/LeaderboardCellResponseDTO";
import { LeaderboardResponseDTO } from "@/core/port/dto/response/leaderboard/LeaderboardResponseDTO";
import { SubmissionWithCodeAndExecutionsResponseDTO } from "@/core/port/dto/response/submission/SubmissionWithCodeAndExecutionsResponseDTO";
import { TicketResponseDTO } from "@/core/port/dto/response/ticket/TicketResponseDTO";

export type AdminDashboardBroadcastRoomCallbacks = {
  ANNOUNCEMENT_CREATED: (announcement: AnnouncementResponseDTO) => void;
  CLARIFICATION_CREATED: (clarification: ClarificationResponseDTO) => void;
  CLARIFICATION_DELETED: (data: { id: string }) => void;
  LEADERBOARD_UPDATED: (leaderboardCell: LeaderboardCellResponseDTO) => void;
  LEADERBOARD_FROZEN: () => void;
  LEADERBOARD_UNFROZEN: (leaderboard: LeaderboardResponseDTO) => void;
  SUBMISSION_CREATED: (
    submission: SubmissionWithCodeAndExecutionsResponseDTO,
  ) => void;
  SUBMISSION_UPDATED: (
    submission: SubmissionWithCodeAndExecutionsResponseDTO,
  ) => void;
  TICKET_CREATED: (ticket: TicketResponseDTO) => void;
  TICKET_UPDATED: (ticket: TicketResponseDTO) => void;
};

export class AdminDashboardBroadcastRoom extends BroadcastRoom<AdminDashboardBroadcastRoomCallbacks> {
  constructor(
    contestId: string,
    public callbacks: AdminDashboardBroadcastRoomCallbacks,
  ) {
    super(`contests/${contestId}/admin`, callbacks);
  }
}

File: applications/backend/core/src/main/kotlin/com/forsetijudge/core/port/driven/broadcast/room/dashboard/AdminDashboardBroadcastRoom.kt
package com.forsetijudge.core.port.driven.broadcast.room.dashboard

import com.forsetijudge.core.domain.entity.Announcement
import com.forsetijudge.core.domain.entity.Clarification
import com.forsetijudge.core.domain.entity.Submission
import com.forsetijudge.core.domain.entity.Ticket
import com.forsetijudge.core.domain.model.Leaderboard
import com.forsetijudge.core.port.driven.broadcast.BroadcastEvent
import com.forsetijudge.core.port.dto.response.announcement.toResponseBodyDTO
import com.forsetijudge.core.port.dto.response.clarification.toIdResponseBodyDTO
import com.forsetijudge.core.port.dto.response.clarification.toResponseBodyDTO
import com.forsetijudge.core.port.dto.response.leaderboard.toResponseBodyDTO
import com.forsetijudge.core.port.dto.response.submission.toWithCodeAndExecutionResponseBodyDTO
import com.forsetijudge.core.port.dto.response.ticket.toResponseBodyDTO
import java.util.UUID

class AdminDashboardBroadcastRoom(
    contestId: UUID,
) {
    companion object {
        val pattern = Regex("/contests/(?<contestId>[a-f0-9\\-]+)/dashboard/admin")
    }

    private val name = "/contests/$contestId/dashboard/admin"

    fun buildAnnouncementCreatedEvent(announcement: Announcement) =
        BroadcastEvent(
            room = name,
            name = "ANNOUNCEMENT_CREATED",
            data = announcement.toResponseBodyDTO(),
        )

    fun buildClarificationCreatedEvent(clarification: Clarification) =
        BroadcastEvent(
            room = name,
            name = "CLARIFICATION_CREATED",
            data = clarification.toResponseBodyDTO(),
        )

    fun buildClarificationDeletedEvent(clarification: Clarification) =
        BroadcastEvent(
            room = name,
            name = "CLARIFICATION_DELETED",
            data = clarification.toIdResponseBodyDTO(),
        )

    fun buildLeaderboardUpdatedEvent(leaderboardCell: Leaderboard.Cell) =
        BroadcastEvent(
            room = name,
            name = "LEADERBOARD_UPDATED",
            data = leaderboardCell.toResponseBodyDTO(),
        )

    fun buildLeaderboardFrozenEvent() =
        BroadcastEvent(
            room = name,
            name = "LEADERBOARD_FROZEN",
        )

    fun buildLeaderboardUnfrozenEvent(leaderboard: Leaderboard) =
        BroadcastEvent(
            room = name,
            name = "LEADERBOARD_UNFROZEN",
            data = leaderboard.toResponseBodyDTO(),
        )

    fun buildSubmissionCreatedEvent(submission: Submission) =
        BroadcastEvent(
            room = name,
            name = "SUBMISSION_CREATED",
            data = submission.toWithCodeAndExecutionResponseBodyDTO(),
        )

    fun buildSubmissionUpdatedEvent(submission: Submission) =
        BroadcastEvent(
            room = name,
            name = "SUBMISSION_UPDATED",
            data = submission.toWithCodeAndExecutionResponseBodyDTO(),
        )

    fun buildTicketCreatedEvent(ticket: Ticket<*>) =
        BroadcastEvent(
            room = name,
            name = "TICKET_CREATED",
            data = ticket.toResponseBodyDTO(),
        )

    fun buildTicketUpdatedEvent(ticket: Ticket<*>) =
        BroadcastEvent(
            room = name,
            name = "TICKET_UPDATED",
            data = ticket.toResponseBodyDTO(),
        )
}
